server.js

```diff
@@ -2,34 +2,237 @@ import express from 'express'
+
+  if (keyword) {
+    const books = await Book.find({ $or: [{ title: keyRegExp }, { authors: keyRegExp }] }).sort(mySort).limit(PAGE_SIZE).skip((page * PAGE_SIZE) - PAGE_SIZE)
+    console.log(keyRegExp)
+    if (books.length > 0) {
+      res.json(books)
+    } else {
+      res.status(404).json({ error: `No books or authors in the database match the keyword '${keyword}'` })
+    }
+
+  } else {
+    const books = await Book.find().sort(mySort).limit(PAGE_SIZE).skip((page * PAGE_SIZE) - PAGE_SIZE)
+    res.json(books)
+  }
+})
+
+app.get('/books/:id', async (req, res) => {
+  const foundBook = await Book.findOne({ bookID: req.params.id })
+  if (foundBook) {
+    res.json(foundBook)
+  } else {
+    res.status(404).json({ error: `No book with id "${req.params.id}" exists.` })
+  }
+})
+
+app.put('/books/:id', async (req, res) => {
+
+  const foundBook = await Book.findOne({ bookID: req.params.id })
+
+  const setRating = (user_rating) => {
+    const totalRating = (foundBook.average_rating * foundBook.ratings_count) + user_rating
+    const totalNumber = foundBook.ratings_count + 1
+    const average = totalRating / totalNumber
+    return Math.round((average + Number.EPSILON) * 100) / 100
+  }
+
+  const updatedBook = await Book.findOneAndUpdate({ bookID: +req.params.id }, {
+    img_url: req.body.img_url ?? foundBook.img_url,
+    average_rating: req.body.user_rating ? setRating(+req.body.user_rating) : foundBook.average_rating,
+    ratings_count: req.body.user_rating ? foundBook.ratings_count + 1 : foundBook.ratings_count
+  }, { new: true })
+  res.json(updatedBook)
+})
+
+app.post('/addbook', async (req, res) => {
+
+  const allBooks = await Book.find()
+  const getLastId = allBooks.sort((a, b) => (a.bookID > b.bookID) ? 1 : -1)
+
+  new Book(
+    {
+      bookID: getLastId[getLastId.length - 1].bookID + 1,
+      title: req.body.title,
+      authors: req.body.author,
+      img_url: req.body.image ?? '',
+      num_pages: req.body.pages ?? 0
+
+    }
+  ).save()
+  res.send('book saved')
+})
+
+//Below is some code I began expirimenting with for my front-end but which I decided to put
+//aside for time being as it would have taken too long to get it all working. (And I'm not sure)
+//if this is really how users would actually be created for a real website. 
+
+app.post('/createuser', async (req, res) => {
+
+  const existingUser = await User.findOne({ username: req.body.username })
+
+  if (!existingUser) {
+    new User(
+      {
+        username: req.body.username,
+        password: req.body.password,
+      }
+    ).save()
+    res.send('User created')
+  } else {
+    res.json({ error: 'That username already exists.' })
+  }
+})
+
+app.post('/login', async (req, res) => {
+  const existingUser = await User.findOne({ password: req.body.password, username: req.body.username })
+  if (existingUser) {
+    res.json(existingUser)
+  } else {
+    res.json({ error: 'Invalid password or username' })
+  }
+
 })
```

Comment:
You should create separate branch to do it, and merge it when it will be finished :)